### PR TITLE
backport upstream hashdb locking changes

### PR DESCRIPTION
**Pull request by magicxyyz**
backports: https://github.com/ethereum/go-ethereum/pull/28542

### CODE REVIEW

File: trie/triedb/hashdb/database.go [hunk 1]
```diff
@@ -183,7 +178,7 @@ func (db *Database) insert(hash common.Hash, node []byte) {
 	db.dirtiesSize += common.StorageSize(common.HashLength + len(node))
 }
 
-// Node retrieves an encoded cached trie node from memory. If it cannot be found
+// node retrieves an encoded cached trie node from memory. If it cannot be found
 // cached, the method queries the persistent database for the content.
 func (db *Database) Node(hash common.Hash) ([]byte, error) {
```
Comment (tsahee) [inline trie/triedb/hashdb/database.go:183]: In this one you did not make the function private. Is that on purpose? Why?

Comment (tsahee) [reply trie/triedb/hashdb/database.go:183]: I'm really curious if / how well we can implement RecordingKV based off pathdb. I don't know the details.
I assume it's not trivial, so I'd say for this PR let's have a Node function exporting node, and leave pathdb support for a separate PR.

Comment (magicxyyz) [reply trie/triedb/hashdb/database.go]: makes sense, I've added the Node method, thanks!

**Review by tsahee — APPROVED** [review]
LGTM

**Pull request merged**